### ISSUE TITLE
[inductor][cpp_wrapper] unused multi-output fallback-kernel outputs are freed only at function return, risking OOM when many are fused into one function

### ISSUE THREAD

**Issue by xpfjmj** [pytorch/pytorch#194602]
### 🐛 Describe the bug

In the `cpp_wrapper` codegen, each output of a multi-output fallback kernel is bound to its own `RAIIAtenTensorHandle`. A handle is only freed at its **last reader** . Outputs that have no downstream reader (e.g. `max_pool3d_with_indices` `indices`) are therefore not freed until the enclosing C++ function returns (their `RAIIAtenTensorHandle` destructor runs only at scope exit). When the whole model is fused into a single `inductor_entry_impl`, these dead outputs stay alive for the entire call and accumulate across every `max_pool3d_with_indices` in the graph, holding multi-GB buffers (e.g. int64 maxpool indices) and causing OOM at large batch. The `py_wrapper` is unaffected because it drops the whole tuple via `del buf` right after keeping the used output.

### A minimal reproducer
Any model that calls a multi-output op whose secondary output is unused (the normal case for `max_pool3d_with_indices`) keeps the unused output handle alive until the enclosing C++ function returns. The bug is codegen-side and device-agnostic (the codegen is shared across backends).

```
import os
import tempfile

os.environ['TORCHINDUCTOR_CPP_WRAPPER'] = '1'
os.environ['TORCHINDUCTOR_CACHE_DIR'] = tempfile.mkdtemp(prefix='repro_mp3d_')

import torch

def f(x):
    out, indices = torch.nn.functional.max_pool3d_with_indices(x, 2, 2)
    y = torch.relu(out)
    return y

x = torch.randn(1, 1, 16, 16, 16, device='cuda')
torch.compile(f, mode='reduce-overhead')(x)
torch.cuda.synchronize()
```

In the generated cpp file, the indices output of maxpool3d (corresponding to `buf2` in the code) is not promptly reset(). Its `RAIIAtenTensorHandle` destructor does not run until inductor_entry_impl exits, so the underlying tensor remains alive throughout the entire function call.
```
void inductor_entry_impl(
    AtenTensorHandle*
        input_handles, // array of input AtenTensorHandle; handles
                        // are stolen; the array itself is borrowed
    AtenTensorHandle*
        output_handles  // array for writing output AtenTensorHandle; handles
                        // will be stolen by the caller; the array itself is
                        // borrowed)
) {
    py::gil_scoped_release release;

    auto inputs = steal_from_raw_handles_to_raii_handles(input_handles, 1);
    auto arg0_1 = std::move(inputs[0]);

    AOTICudaGuard device_guard(0);
    // Topologically Sorted Source Nodes: [max_pool3d_with_indices], Original ATen: [aten.max_pool3d_with_indices]
    AtenTensorHandle buf1_handle;
    AtenTensorHandle buf2_handle;
    AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_cuda_max_pool3d_with_indices(arg0_1, std::array<int64_t, 3>{2L, 2L, 2L}.cbegin(), 3, std::array<int64_t, 3>{2L, 2L, 2L}.cbegin(), 3, std::array<int64_t, 3>{0L, 0L, 0L}.cbegin(), 3, std::array<int64_t, 3>{1L, 1L, 1L}.cbegin(), 3, 0, &buf1_handle, &buf2_handle));
    RAIIAtenTensorHandle buf1(buf1_handle);
    RAIIAtenTensorHandle buf2(buf2_handle);
    arg0_1.reset();

    auto buf3 = std::move(buf1);  // reuse
    // Topologically Sorted Source Nodes: [y], Original ATen: [aten.relu]
    cudaStream_t stream0;
    AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_get_current_cuda_stream(0, (void**)&stream0));
    call_triton_poi_fused_relu_0(buf3, 512L, stream0);
    output_handles[0] = buf3.release();
} // inductor_entry_impl

```

### Versions

PyTorch version: 2.7.1

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by quinnarnold** [comment]
I traced this against current `main`. I used Codex to help inspect the code-generation and lifetime paths; I reviewed the following analysis and agree with it:

> Scheduler DCE records unused multi-outputs in `V.graph.removed_buffers`, but `generate_c_shim_fallback_kernel` still creates an owning `RAIIAtenTensorHandle` for every required return slot. Removed buffers are excluded from the normal free path, so the unused handle receives no `.reset()` and survives until function return. The original max-pool reproduction is now normally decomposed, but the same behavior remains reproducible with a values-only `topk`.

I propose immediately resetting returned `ir.MultiOutput` handles already present in `V.graph.removed_buffers`, with one CPU cpp-wrapper generated-code regression checking numerical correctness and handle placement. Would that scope be welcome, and could the issue be marked `actionable`?

**Comment by xpfjmj** [comment]
Thanks for the analysis! I agree with the proposed scope. Please mark it as actionable. @quinnarnold
